torch/_inductor/comms.py

```diff
@@ -71,6 +74,162 @@ def reorder_compute_for_overlap(
     )
 
 
+def reorder_comms_preserving_peak_memory(
+    snodes: list[BaseSchedulerNode],
+    graph_inputs: OrderedSet[str],
+    graph_outputs: OrderedSet[str],
+) -> list[BaseSchedulerNode]:
+    """
+    Reorders communication ops relative to computation ops to improve communication-compute overlapping and hide comm
+    latency.  Stops moving a particular op if it reaches a point that would have increased the peak memory footprint.
+
+    Currently, follows these heuristics (subject to change or tune):
+    - never reorders collectives relative to one another, for SPMD safety
+    - has an option for per-collective prefetch limit, but does not enable it by default
+    - limits the total number of reorder steps to some factor of the graph size to prevent worst-case quadratic
+      performance
+
+    Prerequisite: sink_comms_and_waits - ensure comm and wait nodes are scheduled as late as possible, respecting data
+    dependencies.  That allows reorder_comms_preserving_peak_memory to take a best case peak-memory snapshot, and then
+    monotonically improve latency by moving collectives backward in time.
+
+    Peak memory impact is computed in an iterative fashion.  First, memory use at each timestep is computed, and global
+    peak memory is computed as a max over timesteps.  Then, when swapping any two adjacent nodes, only the curr-memory
+    for the earlier of the nodes after the swap is affected.  This enables checking step by step whether a swap is
+    peak-memory-safe, and bailing out if not.  Example:
+
+    0   n0      C0
+    1   n1      C0 + Allocs(n1) - Frees(n1)
+    2   n2      C0 + Allocs(n1) - Frees(n1) + Allocs(n2) - Frees(n2)
+
+    0   n0      C0
+    1   n2      C0 + Allocs(n2) - Frees(n2)    <-- After moving n2 to Time 1, only time1 memory changes
+    2   n1      C0 + Allocs(n2) - Frees(n2) + Allocs(n1) - Frees(n1)
```

Comment:
For any node we schedule there are actually two events that affect memory:

- alloc of outputs (affects memory)
- run kernel 
- dealloc of no longer needed tensors (affects memory)

So for n0, n1 there is:

- alloc n0
- dealloc n0 	
- alloc n1
- dealloc n1

The max memory is either:
alloc n0
Or 
alloc n0 - dealloc n0 + alloc n1.

If we swap the two we get:

- alloc n1
- dealloc n1
- alloc n0
- dealloc n0 	

Max memory is either:

alloc n1 
or
alloc n1 - dealloc n1 + alloc n0. 

The difference is (alloc n1 - alloc n0) for the first node and for the second node it's (dealloc n1 - dealloc n0)

But for both nodes, the memory could change.

Reply:
I think that makes sense, but can you articulate your point more clearly?

one thing is that maybe in my comment, i should not be including Frees(n1) on line 1.  I think what I was modeling is 'steady-state memory after running node n1' but that is indeed not the same thing as 'peak memory during n1'.

```
    0   n0      C0
    1   n1      C0 + Allocs(n1) - Frees(n1)
    2   n2      C0 + Allocs(n1) - Frees(n1) + Allocs(n2) - Frees(n2)

```

Apart from that is there anything else you think isn't right?

Reply:
>  I think what I was modeling is 'steady-state memory after running node n1' but that is indeed not the same thing as 'peak memory during n1'.

Don't we want to respect peak memory here, not steady-state memory ?